### PR TITLE
Add support for limit pushdown into connectors

### PR DESCRIPTION
**Pull request by martint**
Add support for limit pushdown in the spirit of https://github.com/prestosql/presto/wiki/Pushdown-of-complex-operations. Also, removed support for legacy table layouts from memory connector and implemented limit pushdown in memory connector.

Depends on https://github.com/prestosql/presto/pull/420

### CODE REVIEW

File: presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java [hunk 1]
```diff
@@ -126,60 +124,63 @@ public synchronized void dropSchema(ConnectorSession session, String schemaName)
...
     }
 
     @Override
     public synchronized Map<String, ColumnHandle> getColumnHandles(ConnectorSession session, ConnectorTableHandle tableHandle)
     {
-        MemoryTableHandle memoryTableHandle = (MemoryTableHandle) tableHandle;
-        return memoryTableHandle.getColumnHandles().stream()
-                .collect(toMap(MemoryColumnHandle::getName, Function.identity()));
+        MemoryTableHandle handle = (MemoryTableHandle) tableHandle;
+        return tables.get(handle.getId())
```
Comment (dain) [inline presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java]: I can't remember... is it ok if this throws an NPE?

File: presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java [hunk 2]
```diff
@@ -1171,6 +1171,20 @@ public boolean usesLegacyTableLayouts(Session session, TableHandle table)
...
+    {
+        ConnectorId connectorId = table.getConnectorId();
+        ConnectorMetadata metadata = getMetadata(session, connectorId);
+
+        if (metadata.usesLegacyTableLayouts()) {
+            return Optional.empty();
+        }
+
+        return metadata.applyLimit(table.getConnectorHandle(), limit)
+                .map(newTable -> new TableHandle(connectorId, newTable, table.getTransaction(), table.getLayout()));
```
Comment (dain) [inline presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java]: If `table.getLayout()` is always null here, maybe just pass null, to keep the call count down.

File: presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java [hunk 3]
```diff
@@ -358,4 +359,17 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
...
         return ImmutableList.copyOf(tables.get(tableId).getDataFragments().values());
     }
+
+    @Override
+    public Optional<ConnectorTableHandle> applyLimit(ConnectorTableHandle handle, long limit)
+    {
+        MemoryTableHandle table = (MemoryTableHandle) handle;
+
+        if (table.getLimit().isPresent() && table.getLimit().getAsLong() < limit) {
+            // can't apply the limit if we already have a lower limit
```
Comment (dain) [inline presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java]: I don't think that is right.  If we are only going to produce 100 rows, and there is a limit of 200, we are still good.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitIntoTableScan.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @Override
+    public Rule.Result apply(LimitNode limit, Captures captures, Rule.Context context)
+    {
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+
+        Optional<TableHandle> newTable = metadata.applyLimit(context.getSession(), tableScan.getTable(), limit.getCount());
+
+        return newTable.map(tableHandle -> Result.ofPlanNode(
```
Comment (dain) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitIntoTableScan.java]: What if I can't completely satisfy the the limit, but I want to update my table handle to reduce overscan?

File: presto-memory/src/main/java/io/prestosql/plugin/memory/TableId.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Objects;
+
+import static java.util.Objects.requireNonNull;
+
+public class TableId
+{
+    private final String connectorId;
```
Comment (dain) [inline presto-memory/src/main/java/io/prestosql/plugin/memory/TableId.java]: Connectors don't really need `connectorId` anymore.  I'd remove this if it isn't too much work.

File: presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java (hunk 1 above)
Comment (martint) [reply presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java]: Yeah, any RuntimeException if tableHandle is invalid is ok

Comment (martint) [reply presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java]: Yeah, any RuntimeException if tableHandle is invalid is ok

File: presto-spi/src/main/java/io/prestosql/spi/connector/LimitApplicationResult.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+package io.prestosql.spi.connector;
+
+public class LimitApplicationResult<T>
+{
+    private final T handle;
+    private final boolean limitGuaranteed;
+
+    public LimitApplicationResult(T handle, boolean limitGuaranteed)
+    {
+        this.handle = handle;
```
Comment (dain) [inline presto-spi/src/main/java/io/prestosql/spi/connector/LimitApplicationResult.java]: add null check

File: presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java [hunk 7]
```diff
@@ -578,4 +578,22 @@ default ConnectorTableProperties getTableProperties(ConnectorSession session, Co
...
+
+    /**
+     * Attempt to push down the provided limit into the table.
+     * <p>
+     * Connectors can indicate whether they don't support limit pushdown or that the action had no effect
+     * by returning {@link Optional#empty()}.
+     * <p>
+     * If the connector could benefit from the information but can't guarantee that it will be able to produce
+     * fewer rows than the provided limit, it should return a non-empty result containing a new handle for the
+     * derived table and the "limit guaranteed" flag set to false.
```
Comment (dain) [inline presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java]: Do we need to mention that it is important for the connector to only return a new handle instance if something changed?

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitIntoTableScan.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Override
+    public Rule.Result apply(LimitNode limit, Captures captures, Rule.Context context)
+    {
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+
+        return metadata.applyLimit(context.getSession(), tableScan.getTable(), limit.getCount())
+                .map(result -> {
+                    PlanNode node = new TableScanNode(
```
Comment (dain) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitIntoTableScan.java]: do we need to check if the handle changed with "equals"?

File: presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java [hunk 9]
```diff
@@ -348,4 +350,16 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
     {
         return ImmutableList.copyOf(tables.get(tableId).getDataFragments().values());
     }
+
+    @Override
+    public Optional<LimitApplicationResult<ConnectorTableHandle>> applyLimit(ConnectorTableHandle handle, long limit)
+    {
+        MemoryTableHandle table = (MemoryTableHandle) handle;
+
+        if (!table.getLimit().isPresent() || limit < table.getLimit().getAsLong()) {
```
Comment (dain) [inline presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java]: you could use `table.getLimit().orElse(Long.MAX_VALUE)`

File: presto-memory/src/main/java/io/prestosql/plugin/memory/ColumnInfo.java [hunk 10]
```diff
@@ -13,42 +13,37 @@
...
 
-import java.util.Objects;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.connector.ColumnMetadata;
+import io.prestosql.spi.type.Type;
 
 import static java.util.Objects.requireNonNull;
 
-public final class MemoryConnectorId
+public class ColumnInfo
```
Comment (dain) [inline presto-memory/src/main/java/io/prestosql/plugin/memory/ColumnInfo.java]: Maybe add `toString`

**Review by dain — APPROVED** [review]
Couple of questions and suggestions.  If no major changes are needed, looks good to me.

File: presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java (hunk 7 above)
Comment (martint) [reply presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java]: It's there, but probably not clear enough. I'll make more emphasis on that.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitIntoTableScan.java (hunk 8 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitIntoTableScan.java]: That's not necessary and probably not desirable. A connector may use handles as pure identifiers and record the limit in some sidecar data structure, so it'd need to return the same handle.

**Comment by findepi** [comment]
🚀

**Pull request merged**